File: Projects/SOL_Python (2026-09-04)/SOL_Python/geod_to_geoc.m
function [phi_geoc, h_geod] = geod_to_geoc(phi_geod, a,b, h0)
% Retrieves geocentric latitude φ and geodetic altitude h_ϕ (above a) for the geodetic latitude ϕ
% h_geod ≠ R_e
% [phi_gc, h_gd] = geod_to_geoc(60.4350632585099, Earth.r1_km, 0.8 * Earth.r1_km) == [50, 831.969398327804]
% [phi_gc, h_gd] = geod_to_geoc([0, 90], a,b) == [0, a-b]
%{
% Test case:
    phi_geod = 50;
    h0 = 1000;
    [phi_geoc, h_geod] = geod_to_geoc(phi_geod, Earth.r1_km,Earth.r2_km, h0)
    % phi_geoc = 49.8364099440923°
    % h_geod = 1012.51026593283 km

    % → identical the Matlab Aerospace Toolbox function → using h_φ
    [phi_geoc, radii] = geod2geoc(phi_geod, h_geod, Earth.f,Earth.r1_km)
    % phi_geoc = 49.8364099440923°
    % radii - Earth.r1_km == h0

    [phi_geoc, radii] = geod2geoc(phi_geod, h0, Earth.f,Earth.r1_km)
    % phi_geoc = 49.8361320923381°
    % radii - Earth.r1_km = 987.489785146104 < h0 OK
%}
    if nargin < 4
        h0 = 0;
    end

    rho = a + h0;

    if 0  % Derivation #1 (for h0 = 0)
        %{
            Using identity for simplifying:
           ┌────────────────────────────┐ 
           │ 1 - e² sin(ϕ)² ≡ a² / R_N² │
           └────────────────────────────┘ 
        
            % Based on equation along the circle encompassing the ellipsoid
            x = (R_N + h_geod) * cosd(phi_geod)  % == x_sat 
            z = (R_N * (1 - e2) + h_geod) * sind(phi_geod)  % == z_sat
            r = hypot(x, z)  % == a
            [ r, norm(R_sat) , a ]
            
            h = h_geod  % (sought solution)
            a^2 == (R_N + h)^2 * cosd(phi_geod)^2 + (R_N * (1 - e2) + h)^2 * sind(phi_geod)^2
    
            % → isolate h
        %}
        e2 = 1 - (b/a)^2;
        R_N = a ./ sqrt(1 - e2 * sind(phi_geod).^2);

        % Quadratic form solution
        B = rho^2 ./ R_N;
        C = -rho^2 + R_N.^2 .* (2 * rho^2 ./ R_N.^2 + e2^2 * sind(phi_geod).^2 - 1);
        % (maybe this can still be simplified?)
        h_geod = sqrt(B.^2 - C) - B;

        % Once we know h_ϕ:
        [R_sat, phi_geoc] = geod_to_pos(phi_geod, 0, h_geod, a,b);
    
    else  % Derivation #2 (OK for any h0)

        % a) Computes sub-satellite point on the ellipsoid surface
        [R_sub, ~, ~, R_N] = geod_to_pos(phi_geod, 0, 0, a,b);  % (h = 0)
        x_sub = R_sub(1, :);
        z_sub = R_sub(3, :);

        % b) Computes normal line perpendicular to ellipsoid at sub-satellite point
        if 0
            m = b * -x_sub ./ (a^2 * sqrt(1 - (x_sub / a).^2));  % tangential slope to sub-sat point
            m = m .* sign(phi_geod);  % handles negative ϕ
            m = -1 ./ m;  % perpendicular
        else
            m = cotd(90 - phi_geod); % (direct solution)
        end
        b_ = z_sub - m .* x_sub;  % ordinate at origin ≠ ellipsoid parameter b
 
        % c) Intersecting normal vector at R_sub with spherical Earth @ ρ
        % m * xx + b_normal == sqrt(a_^2 - xx^2) @ x = x_sat
        % [ m^2*xx^2 + 2*b_normal*m*xx + b_normal^2 , a^2 - xx^2 ]
        % [ (m^2+1)*xx^2 + 2*b_normal*m*xx + b_normal^2 - a^2, 0 ]
        if 0
            A = m.^2 + 1;
            B = 2 * b_ .* m;
            C = b_.^2 - rho.^2;
            x_sat = (-B + sqrt(B.^2 - 4 * A .* C)) ./ (2*A);  % Quadratic equation solution
        else
            x_sat = (-b_ .* m + sqrt(rho.^2 * (m.^2 + 1) - b_.^2)) ./ (m.^2 + 1);  % other form
        end

        z_sat = m .* x_sat + b_;

        h_geod = x_sat ./ cosd(phi_geod) - R_N;  % h_ϕ (Vermeille's algoithm)
        phi_geoc = atan2d(z_sat, x_sat);  % φ

        % Handles NaNs
        k = ismember(phi_geod, [-90, 0, 90]);
        phi_geoc(k) = phi_geod(k);
        h_geod(phi_geod == 0) = h0;
        h_geod(abs(phi_geod) == 90) = a + h0 - b;
    end
    % could also be solved in R,ϕ :
    % y = R sin(ϕ) + z₀ ; y = sqrt((a+h0)² - R² cos(ϕ)²)
    % R^2 c1 + c2 == sqrt(c3 - R^2 c4)
    % (R^2 c1 + c2)^2 == c3 - R^2 c4
    % ... quadratic solve for R

    % ==> ! verify ± solution depending on quadrants...

end  % function

File: Projects/SOL_Python (2026-09-04)/SOL_Python/geod_to_geoc.py
"""
geod_to_geoc.py

Retrieves the geocentric latitude phi and the geodetic-direction altitude
h_phi (above a) for a given geodetic latitude phi_geod.

Note: h_geod returned here is != R_e (the Earth radius); see Vermeille's
algorithm reference below.

Examples (from the original MATLAB docstring)
-----------------------------------------------
    geod_to_geoc(60.4350632585099, Earth.r1_km, 0.8 * Earth.r1_km)
        -> (phi_geoc, h_geod) == (50, 831.969398327804)

    geod_to_geoc([0, 90], a, b)
        -> (phi_geoc, h_geod) == ([0, 90], [0, a - b])

Translated from the original MATLAB implementation. Only "Derivation #2"
is implemented, since "Derivation #1" was inside an `if 0` (dead-code /
reference-only) branch in the original source; it is kept below purely as
documentation.

Derivation #1 (reference only, not executed, valid only for h0 = 0)
---------------------------------------------------------------------
Using the identity  1 - e^2 sin(phi)^2 == a^2 / R_N^2  to simplify:

    x = (R_N + h_geod) * cos(phi_geod)                  # == x_sat
    z = (R_N * (1 - e2) + h_geod) * sin(phi_geod)        # == z_sat
    r = hypot(x, z)                                      # == a
    # [r, norm(R_sat), a] should all match

    h = h_geod  (sought solution), solving:
    a^2 == (R_N + h)^2 * cos(phi_geod)^2
         + (R_N * (1 - e2) + h)^2 * sin(phi_geod)^2
    for h via the quadratic form (B, C below), then:
    [R_sat, phi_geoc] = geod_to_pos(phi_geod, 0, h_geod, a, b)
"""

import numpy as np
from SOL_Tools.AstroConstants import *  # all variales
from SOL_Tools.Math_tools import * 

from geod_to_pos import geod_to_pos

def geod_to_geoc(phi_geod, a=6378.137, b=6356.75231424518, h0=0.0):
    """
    Parameters
    ----------
    phi_geod : array_like
        Geodetic latitude(s) [deg].
    a : float, optional
        Semi-major axis. Defaults to WGS84 Earth value (km).
    b : float, optional
        Semi-minor axis. Defaults to WGS84 Earth value (km).
    h0 : float, optional
        Altitude above `a` (i.e. rho = a + h0 is the radius of the
        reference sphere onto which the geodetic-normal line is
        projected). Default = 0.

    Returns
    -------
    phi_geoc : ndarray
        Geocentric latitude(s) [deg].
    h_geod : ndarray
        Geodetic-direction altitude(s) h_phi (Vermeille's algorithm).
        Note h_geod != R_e.
    """

    phi_geod = np.atleast_1d(np.asarray(phi_geod, dtype=float))
    rho = a + h0

    # --- Derivation #2 (valid for any h0) ----------------------------------

    # a) Sub-satellite point on the ellipsoid surface (h = 0)
    R_sub, _, _, R_N, *_ = geod_to_pos(phi_geod, 0.0, 0.0, a, b)
    x_sub = R_sub[0, :]
    z_sub = R_sub[2, :]

    # b) Normal line perpendicular to the ellipsoid at the sub-satellite point
    #    (direct solution, equivalent to the tangential-slope derivation
    #    commented out in the original MATLAB code)
    m = cotd(90 - phi_geod)
    b_ = z_sub - m * x_sub  # y-intercept (unrelated to the ellipsoid parameter b)

    # c) Intersect the normal line at R_sub with the sphere of radius rho
    #    m*xx + b_ == sqrt(rho^2 - xx^2)  @ x = x_sat
    x_sat = (-b_ * m + np.sqrt(rho**2 * (m**2 + 1) - b_**2)) / (m**2 + 1)
    z_sat = m * x_sat + b_

    h_geod = x_sat / cosd(phi_geod) - R_N          # h_phi (Vermeille's algorithm)
    phi_geoc = atan2d(z_sat, x_sat)                 # phi

    # Handle special/singular cases (poles and equator)
    k = np.isin(phi_geod, [-90.0, 0.0, 90.0])
    phi_geoc = np.where(k, phi_geod, phi_geoc)
    h_geod = np.where(phi_geod == 0.0, h0, h_geod)
    h_geod = np.where(np.abs(phi_geod) == 90.0, a + h0 - b, h_geod)

    # NOTE (from original source): could also be solved in (R, phi):
    #   y = R*sin(phi) + z0 ; y = sqrt((a+h0)^2 - R^2*cos(phi)^2)
    #   R^2*c1 + c2 == sqrt(c3 - R^2*c4)
    #   (R^2*c1 + c2)^2 == c3 - R^2*c4   -> quadratic solve for R
    # ==> verify +/- solution depending on quadrant, as in the original.

    return phi_geoc, h_geod


if __name__ == "__main__":
    # Test case from the original MATLAB docstring
    phi_geod = 50.0
    h0 = 1000.0

    a = 6378.137
    b = 6356.7523142

    phi_geoc, h_geod = geod_to_geoc(phi_geod, a, b, h0)
    print(f"phi_geoc = {phi_geoc[0]:.10f}  (expected ~49.8364099440923)")
    print(f"h_geod   = {h_geod[0]:.8f}     (expected ~1012.51026593283)")

    # Second documented example
    phi_geoc2, h_geod2 = geod_to_geoc(60.4350632585099, a, 0.8 * a)
    print(f"\nphi_geoc = {phi_geoc2[0]:.6f}  (expected 50)")
    print(f"h_geod   = {h_geod2[0]:.6f}  (expected 831.969398327804)")

    # Poles / equator edge cases
    phi_geoc3, h_geod3 = geod_to_geoc([0.0, 90.0], a, b)
    print(f"\nphi_geoc = {phi_geoc3}  (expected [0, 90])")
    print(f"h_geod   = {h_geod3}  (expected [0, a-b] = [0, {a - b}])")